File: Unity/Assets/PixelPlanets/StellarBreaker/Scripts/Gameplay/ArtifactService.cs
using System;
using System.Collections.Generic;
using StellarBreaker.Core;
using StellarBreaker.Config;
using StellarBreaker.Economy;

namespace StellarBreaker.Gameplay
{
    /// <summary>
    /// Permanent artifacts bought with Relics. Each adds a multiplicative bonus to a stat.
    /// Survives prestige (Relics + artifact levels are not reset).
    /// </summary>
    public class ArtifactService
    {
        readonly IReadOnlyList<ArtifactDefinition> _defs;
        readonly CurrencyService _relics;
        readonly int[] _levels;

        public event Action<int, int> OnArtifactChanged;

        public ArtifactService(IReadOnlyList<ArtifactDefinition> defs, CurrencyService relicWallet)
        {
            _defs   = defs ?? throw new ArgumentNullException(nameof(defs));
            _relics = relicWallet ?? throw new ArgumentNullException(nameof(relicWallet));
            _levels = new int[_defs.Count];
        }

        public int Count => _defs.Count;
        public int LevelOf(int i) => _levels[i];
        public ArtifactDefinition Def(int i) => _defs[i];

        /// <summary>Restore artifact levels from a save (extra/short entries ignored).</summary>
        public void RestoreLevels(int[] levels)
        {
            if (levels == null) return;
            int n = System.Math.Min(levels.Length, _levels.Length);
            for (int i = 0; i < n; i++) _levels[i] = levels[i] < 0 ? 0 : levels[i];
        }

        public BigNumber NextCost(int i)
            => UpgradeCost.Exponential(_levels[i] + 1, _defs[i].baseCost, _defs[i].costGrowth);

        public bool BuyOrUpgrade(int i)
        {
            if (!_relics.TrySpend(NextCost(i))) return false;
            _levels[i]++;
            OnArtifactChanged?.Invoke(i, _levels[i]);
            return true;
        }

        /// <summary>Upgrade artifact i as many times as Relics allow (real cost each step →
        /// cannot overspend). Returns levels bought. Capped for safety.</summary>
        public int BuyOrUpgradeMax(int i, int cap = 100000)
        {
            int n = 0;
            while (n < cap && BuyOrUpgrade(i)) n++;
            return n;
        }

        /// <summary>Current fractional bonus for artifact i (0 if unowned). UI-friendly.</summary>
        public double LevelBonus(int i) => _defs[i].BonusAt(_levels[i]);

        /// <summary>Aggregate multiplier for a stat = ∏ (1 + bonus(level)).</summary>
        public BigNumber Multiplier(ArtifactEffect effect)
        {
            var mult = BigNumber.One;
            for (int i = 0; i < _defs.Count; i++)
            {
                if (_defs[i].effect != effect || _levels[i] <= 0) continue;
                mult = mult * new BigNumber(1.0 + _defs[i].BonusAt(_levels[i]));
            }
            return mult;
        }

        public BigNumber DpsMultiplier()       => Multiplier(ArtifactEffect.Dps);
        public BigNumber GoldMultiplier()      => Multiplier(ArtifactEffect.Gold);
        public BigNumber TapDamageMultiplier() => Multiplier(ArtifactEffect.TapDamage);
    }
}

File: TelegramApp/src/game/gameplay/ArtifactService.ts
// Ported from Assets/PixelPlanets/StellarBreaker/Scripts/Gameplay/ArtifactService.cs
import { BigNumber } from '../core/BigNumber'
import { Emitter } from '../core/Emitter'
import { ArtifactEffect, artifactBonusAt, type ArtifactDefinition } from '../config/ArtifactDefinition'
import type { CurrencyService } from '../economy/CurrencyService'
import { upgradeCostExponential } from '../economy/UpgradeCost'

/**
 * Permanent artifacts bought with Relics. Each adds a multiplicative bonus to a stat.
 * Survives prestige (Relics + artifact levels are not reset).
 */
export class ArtifactService {
  private readonly defs: ArtifactDefinition[]
  private readonly relics: CurrencyService
  private readonly levels: number[]

  readonly onArtifactChanged = new Emitter<{ index: number; level: number }>()

  constructor(defs: ArtifactDefinition[], relicWallet: CurrencyService) {
    this.defs = defs
    this.relics = relicWallet
    this.levels = new Array(defs.length).fill(0)
  }

  get count(): number {
    return this.defs.length
  }
  levelOf(i: number): number {
    return this.levels[i]
  }
  def(i: number): ArtifactDefinition {
    return this.defs[i]
  }

  /** Restore artifact levels from a save (extra/short entries ignored). */
  restoreLevels(levels: number[] | undefined | null): void {
    if (!levels) return
    const n = Math.min(levels.length, this.levels.length)
    for (let i = 0; i < n; i++) this.levels[i] = levels[i] < 0 ? 0 : levels[i]
  }

  nextCost(i: number): BigNumber {
    return upgradeCostExponential(this.levels[i] + 1, this.defs[i].baseCost, this.defs[i].costGrowth)
  }

  buyOrUpgrade(i: number): boolean {
    if (!this.relics.trySpend(this.nextCost(i))) return false
    this.levels[i]++
    this.onArtifactChanged.emit({ index: i, level: this.levels[i] })
    return true
  }

  /**
   * Upgrade artifact i as many times as Relics allow (real cost each step ->
   * cannot overspend). Returns levels bought. Capped for safety.
   */
  buyOrUpgradeMax(i: number, cap = 100_000): number {
    let n = 0
    while (n < cap && this.buyOrUpgrade(i)) n++
    return n
  }

  /** Current fractional bonus for artifact i (0 if unowned). UI-friendly. */
  levelBonus(i: number): number {
    return artifactBonusAt(this.defs[i], this.levels[i])
  }

  /** Aggregate multiplier for a stat = prod (1 + bonus(level)). */
  multiplier(effect: ArtifactEffect): BigNumber {
    let mult = BigNumber.One
    for (let i = 0; i < this.defs.length; i++) {
      if (this.defs[i].effect !== effect || this.levels[i] <= 0) continue
      mult = mult.mul(new BigNumber(1 + artifactBonusAt(this.defs[i], this.levels[i])))
    }
    return mult
  }

  dpsMultiplier(): BigNumber {
    return this.multiplier(ArtifactEffect.Dps)
  }
  goldMultiplier(): BigNumber {
    return this.multiplier(ArtifactEffect.Gold)
  }
  tapDamageMultiplier(): BigNumber {
    return this.multiplier(ArtifactEffect.TapDamage)
  }
}
